### PR TITLE
OpenStack Federation blog article

### PR DESCRIPTION
**Pull request by miguelgrinberg**
Bio: Miguel Grinberg is a software engineer with a background in web technologies and REST APIs. He is the author of the book "Flask Web Development" from O'Reilly Media, and has a blog at http://blog.miguelgrinberg.com, where he writes about a variety of topics including web development, robotics, photography and the occasional movie review. Miguel works as a Software Developer with the Rackspace Private Cloud team. He lives in Portland, Oregon with his wife, four kids, two dogs and a cat. Follow @miguelgrinberg on Twitter.

Twitter: @miguelgrinberg
GitHub: https://github.com/miguelgrinberg
LinkedIn: https://www.linkedin.com/in/miguelgrinberg

Redaction, editing or structure changes by technical writer staff: Yes

Desired date: 2015-09-16

### CODE REVIEW

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+    - openstack
+    - keystone
+    - osad
+    - federation
+---
+
+# Keystone-to-Keystone Federation with the openstack-ansible Project
+
+Federation support in OpenStack has been greatly improved in the Kilo release,
+and while there are still some rough edges, the feature is now usable, if you
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: and, while there are ...

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+don't mind adding a little elbow grease. This article focuses on one specific
+configuration called *Keystone-to-Keystone federation* (or K2K), in which users
+of an OpenStack cloud use their credentials to access services in another
+OpenStack cloud, with both clouds using Keystone as their identity services.
+
+<!-- more -->
+
+The purpose of this article is two-fold. First, I want to show you how to
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: ... is twofold. First, I ...

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 3]
```diff
@@ -0,0 +1,554 @@
...
+Then, I want to touch on some of those rough edges I mentioned above, so that
+you have your expectations set at the right level.
+
+## What is OpenStack Federation?
+
+In a standard, non-federated set up, users authenticate against the cloud's
+own identity APIs, usually implemented by the Keystone service. With Keystone
+federation, the options for authentication are expanded, by creating a
+separation between the entity that authenticates the user (the *identity
+provider*, or IdP), and the entity that owns the cloud resources the user
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: ... or IdP) and the entity ....

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+The purpose of this article is two-fold. First, I want to show you how to
+configure K2K federation between two [openstack-ansible](https://github.com/openstack/openstack-ansible) clouds.
+Then, I want to touch on some of those rough edges I mentioned above, so that
+you have your expectations set at the right level.
+
+## What is OpenStack Federation?
+
+In a standard, non-federated set up, users authenticate against the cloud's
+own identity APIs, usually implemented by the Keystone service. With Keystone
+federation, the options for authentication are expanded, by creating a
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: ... are expanded by creating ...

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+separation between the entity that authenticates the user (the *identity
+provider*, or IdP), and the entity that owns the cloud resources the user
+wants to access (the *service provider*, or SP).
+
+Since we are talking about OpenStack, it is obvious that the service provider
+in all the federation use cases is an OpenStack cloud. But on the IdP side
+things are more interesting, because it is now possible to use a variety of
+external identity services, such as those based on ADFS, OpenID Connect or
+SAML2. The end result is that an external IdP service that has a trust
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: SAML2. As a result, an external ...

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+from other OpenStack distributions, is that the OpenStack services are
+installed in individual LXC containers, so they are isolated from each other.
+Configuration options are given in [YAML](http://yaml.org/) files, which are
+simple, hierarchical text files that can be edited in your favorite text
+editor. The Ansible model simplifies the management of the OpenStack cluster,
+so that for example, adding a new compute node to an existing cloud just
+requires editing the configuration file that lists all the nodes, and then
+re-running the playbooks. Each time you run the playbooks, Ansible detects what
+changed in the configuration files, and invokes the appropriate tasks to
+update the cluster accordingly, with minimal or no disruption to the cluster.
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: ... accordingly with minimal ...

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 7]
```diff
@@ -0,0 +1,554 @@
...
+If one of the services starts to misbehave, just delete its container and run
+the Ansible playbooks to have a brand new container be created to takes the
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: ... playbooks to create a brand new container to take the

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 8]
```diff
@@ -0,0 +1,554 @@
...
+    aio1_swift_proxy_container-0f90e545           RUNNING  10.0.3.146, 172.29.238.15, 172.29.245.109  -     YES (onboot, openstack)
+    aio1_utility_container-d7a0a0e0               RUNNING  10.0.3.125, 172.29.239.67                  -     YES (onboot, openstack)
+
+The first interesting thing you may notice is that even though this is a
+single-node deployment, the use of containers makes it possible to deploy
+redundant services. You can see in the above listing that keystone, rabbitmq,
+galera and a few others are deployed in multiple containers. The set up script
+installs haproxy as a load balancer for these.
+
+Among all the containers you'll find one with the name `utility`, usually at
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: ... the containers, you'll find ...

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 9]
```diff
@@ -0,0 +1,554 @@
...
+the bottom of the list. This is a container that has all the OpenStack command
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: ... This container has all ...

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 10]
```diff
@@ -0,0 +1,554 @@
...
+    | 595b2db82173460a994d23b29967a207 | glance                         |
+    | 5fa7bf9a79aa4782b5657bd23d95082c | heat                           |
+    | 8eb91912e1e2466ab781b317c358c109 | neutron                        |
+    | 925e531fdbf340bbbcd0825d9c6f94f1 | keystone                       |
+    | af7ecc9999c44e6e8459ddc3810e128d | ceilometer                     |
+    | c9c2e4e7fa264c10809121ffda2a9d14 | admin                          |
+    | fedf070e403343d18fadb04dbc063048 | nova                           |
+    +----------------------------------+--------------------------------+
+
+In the following sections I will describe how to connect two openstack-ansible
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: ... following sections, I will ...

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 11]
```diff
@@ -0,0 +1,554 @@
...
+these settings there are a few that configure the SSL certificate that is used
+by the SAML2 protocol, which is automatically generated as part of the install.
+Out of all these variables, the `regen_cert` variable deserves a mention. This
+variable can be used to force Ansible to replace an existing certificate with
+a new one, simply by setting the variable to `true`. If this variable is set
+to `false`, then repeated runs of the playbook will leave the same certificate
+installed.
+
+Are you ready to convert one of your two test clouds to an identity provider?
+Just add the above snippet to `/etc/openstack_deploy/user_variables.yml` and
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: .. to `/etc/openstack_deploy/user_variables.yml`, and

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 12]
```diff
@@ -0,0 +1,554 @@
...
+make sure you write the IP address or hostname of your other cloud in the
+`keystone_sp_host` variable. To make this configuration change effective you
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: ... change effective, you

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 13]
```diff
@@ -0,0 +1,554 @@
...
+just simply need to run the playbooks again. Since this change only affects
+keystone, we will just run the keystone playbook to save a bit of time:
+
+    # cd /opt/openstack-ansible/playbooks
+    # openstack-ansible os-keystone-install.yml
+
+Ansible will now run for a few minutes, and when the playbook completes your
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: ... and, when the playbook completes, your

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 14]
```diff
@@ -0,0 +1,554 @@
...
+                    local:
+                      - group:
+                          name: fedgroup
+                          domain:
+                            name: Default
+
+Here the `cert_duration_years` configures the duration of a self-signed SSL
+certificate that will protect the endpoints used by the SAML2 protocol. The
+`trusted_idp_list` variable is the SP counterpart of the `service_providers`
+variable in the IdP configuration, and defines the list of identity providers.
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: ... configuration and defines ...

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 15]
```diff
@@ -0,0 +1,554 @@
...
+
+Like we did for SPs in the IdP side, here we have to provide a few settings for
+each IdP. These include the IdP's unique identifier and its endpoints, plus
+a location where metadata associated with the IdP is to be stored. These can be
+set using custom variables.
+
+The `federated_identities` defines sets of identities in the SP cloud that will
+be used with remote users. These can include a domain, project, group, role,
+and/or user. This is one part of the configuration that needs to be customized
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: or user. ...

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 16]
```diff
@@ -0,0 +1,554 @@
...
+SP can use to determine how to handle the user.
+
+The protocol's `mapping` section defines how the users of the IdP are to be
+mapped to local users and this is done through a list of rules. When the
+Keystone service running on the SP cloud receives a remote authentication,
+it will evaluate the `remote` portion of each rule until one matches the user.
+The matching criteria can be pretty simple or can also be fairly complex,
+involving one or more of the declared attributes. Once a matching rule has been
+found, the user is mapped as a local user according to the `local` part of the
+rule. In the example above there is a single rule, which says that the remote
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: ...  example above, there is ...

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 17]
```diff
@@ -0,0 +1,554 @@
...
+user must have a declared `openstack_user` value (the username). The user will
+be mapped to the group `fedgroup` in the `Default` domain. Since a username
+isn't given in the `local` part of the rule, the remote user will be an
+ephemeral user that will cease to exist once the user leaves.
+
+The rule definition language allows for rules much more complex than the one
+shown above. For more information about federation mappings you can consult the
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: ... federation mappings, you can ...

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 18]
```diff
@@ -0,0 +1,554 @@
...
+[appropriate section of the Keystone documentation](http://docs.openstack.org/developer/keystone/mapping_combinations.html).
+
+To configure your second cloud as a SP, add the above configuration to your
+`/etc/openstack_deploy/user_variables.yml` file and, like in the IdP case,
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: ... file, and, like ...

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 19]
```diff
@@ -0,0 +1,554 @@
...
+run the Keystone playbook:
+
+    # cd /opt/openstack-ansible/playbooks
+    # openstack-ansible os-keystone-install.yml
+
+At this point the K2K federation between your two clouds should be fully
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: At this point, the K2K ...

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 20]
```diff
@@ -0,0 +1,554 @@
...
+
+## The K2K Federation Authentication Flow
+
+If you followed the instructions above, you have one of your clouds configured
+as an IdP, and the other configured as a SP. Not only that, the Keystone
+playbook in openstack-ansible also set things up so that the IdP cloud knows
+about the SP cloud, and the SP cloud knows about the IdP cloud.
+
+So how does federation work? To begin, the user must authenticate against the
+IdP cloud. This is the standard authentication, so for command line usage this
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: ... so, for command line usage, this ...

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 21]
```diff
@@ -0,0 +1,554 @@
...
+    # OBJECT_STORE_URL=http://23.253.97.96:8080/v1/AUTH_bf7e5af2a9a24309bf3b10edad12ee71
+    # ORCHESTRATION_URL=http://23.253.97.96:8004/v1/bf7e5af2a9a24309bf3b10edad12ee71
+    # VOLUME_URL=http://23.253.97.96:8776/v1/bf7e5af2a9a24309bf3b10edad12ee71
+    #----------------------------------------
+    # OpenStack client setup:
+    export OS_TOKEN=<scoped-token>
+    export OS_URL=<desired-service-endpoint>
+
+The output of the `federated-login.sh` shows a progress of all the actions
+taken during the authentication process, and once the authentication is
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: ... authentication process and, once the ...

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 22]
```diff
@@ -0,0 +1,554 @@
...
+completed, displays the information that you need to access the SP cloud.
+
+From the output of the script, the important bits are the scoped token and
+the URLs for all the services. The token is going to be used with all the
+services, so it is best to put it in an environment variable that the openstack
+command line client automatically picks up, as suggested by the script:
+
+    # export OS_TOKEN=<scoped-token>
+
+Then for example, if you wanted to obtain a listing of all the images available
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: Then, for ...

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 23]
```diff
@@ -0,0 +1,554 @@
...
+service endpoint for each command, given that when accessing the local cloud
+the openstack command line client can figure that out on its own from the
+service catalog. I will discuss this and other problems with the current state
+of K2K federation in the next section.
+
+## Current Limitations of K2K
+
+Now that you've seen, and maybe even experienced, K2K federation, I'm sure you
+are scratching your head and wondering why things aren't more streamlined. In
+this section I have compiled a short list of problems you may run into when
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: this section, I have compiled ...

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 24]
```diff
@@ -0,0 +1,554 @@
...
+unfortunate answer is no, because the implementation of the SAML2 federation
+protocol in Keystone is incomplete. The SP side has support for the "websso"
+flow, which enables Horizon to connect to an external IdP, but the Keystone
+IdP implementation does not currently support websso, restricting Horizon to
+only connect to third party IdPs at this time. An example IdP that works well
+with Horizon is the Active Directory Federation Service from Microsoft (ADFS).
+
+### Command Line Client Support
+
+In the previous section you saw that to use the `openstack` command line client
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: ... previous section, you saw ...

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 25]
```diff
@@ -0,0 +1,554 @@
...
+happens transparently during the authentication process, the command line
+clients know how to do all this for you.
+
+At this time, the authentication flow for K2K is not fully integrated
+with the OpenStack clients, so a fully automated authentication similar to that
+of a local cloud is not possible. Work on a K2K federation authentication
+extension for Keystone is currently in progress, so this limitation is likely
+going to be a non-issue in the Liberty release. My hope is that with some
+additional variables added to the user's `openrc` file, the command line
+clients can perform the complete federated authentication, similar to what the
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: ... similar to the

File: _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md [hunk 26]
```diff
@@ -0,0 +1,554 @@
...
+hope I said enough good things about it that you will decide to check it out in
+detail, even if you don't plan to use federation.
+
+My intention is to write a follow up federation article when the Liberty
+release ships. I hope by that time some of the problems I enumerated in this
+article will have a solution.
+
+Thank you!
+
+Miguel
```
Comment (catlook) [inline _posts/2015-09-16-keystone-to-keystone-federation-with-openstack-ansible.md]: [Editor's note: Thanks, Miguel!  Feel free to use any of my edits that make sense and discard the rest!  -- Cat]

**Comment by miguelgrinberg** [comment]
@catlook thanks a lot for your feedback! Your edits are now incorporated into the article.

@rgbkrk is the build error caused by my article somehow? The stack trace isn't terribly clear, seems unrelated to me.

I'll give my team mates a chance to review the final text tomorrow. I think this is in a good shape to be published on Wednesday.

**Comment by ktbartholomew** [comment]
@miguelgrinberg the failed build is not your fault.

**Comment by rgbkrk** [comment]
Build is a go now. Still want your team mates to review, or do you want to :shipit:?

**Comment by miguelgrinberg** [comment]
@rgbkrk I think I have one small change to make, I'll let you know when it is ready to go. Thanks!

**Comment by miguelgrinberg** [comment]
@rgbkrk Okay, this is now ready to go online. Thanks!

**Pull request merged**